### PR TITLE
New module lightsail_lb for AWS Lightsail Load Balancer

### PR DESCRIPTION
**Pull request by gencebay**
##### SUMMARY
Creates or deletes load balancer and attaches running instances to the created load balancer in AWS Lightsail.

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
ansible/modules/cloud/amazon/lightsail_lb.py

### CODE REVIEW

**Comment by gencebay** [comment]
bot_status

**Comment by gencebay** [comment]
bot_broken

**Comment by jctanner** [comment]
bot_status

**Comment by jctanner** [comment]
@gencebay sorry about that. The bot was having issues with a certain jinja test filter. All fixed now.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+except ImportError:
+    HAS_BOTOCORE = False
+
+try:
+    import boto3
+except ImportError:
+    # will be caught by imported HAS_BOTO3
+    pass
+
+from ansible.module_utils.basic import AnsibleModule
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py]: New AWS modules should use AnsibleAWSModule, which drastically reduces the amount of boilerplate needed. See https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#porting-code-to-ansibleawsmodule

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+    if not region:
+        module.fail_json(msg='region must be specified')
+
+    client = None
+    try:
+        client = boto3_conn(module, conn_type='client', resource='lightsail',
+                            region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except (botocore.exceptions.ClientError, botocore.exceptions.ValidationError) as e:
+        module.fail_json(msg='Failed while connecting to the lightsail service: %s' % e, exception=traceback.format_exc())
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py]: With AnsibleAWSModule everything up to here can be replaced by `client = module.client('lightsail')`

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        healthCheckPath=dict(type='str', default=None)
+    ))
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='Python module "boto3" is missing, please install it')
+
+    if not HAS_BOTOCORE:
+        module.fail_json(msg='Python module "botocore" is missing, please install it')
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py]: These checks are handled by AnsibleAWSModule.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+    try:
+        core(module)
+    except (botocore.exceptions.ClientError, Exception) as e:
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py:296]: It's rarely appropriate to catch all Exceptions.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        module.fail_json(msg=str(e), exception=traceback.format_exc())
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py]: `module.fail_json_aws(e)`

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+changed:
+  description: if a load balancer has been modified/created
+  returned: always
+  type: bool
+  sample:
+    changed: true
+loadbalancer:
+  description: load balancer info
+  returned: when state is present
+  type: dict
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py:90]: This should be more specific about the contents of the data structure that's returned.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+  state:
+    description:
+      - Indicate desired state of the target.
+    default: present
+    choices: ['present', 'absent']
+  name:
+    description:
+      - The name of the load balancer.
+    required: true
+  region:
```
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/lightsail_lb.py]: `region` is part of the standard ec2 documentation fragment and doesn't need to be re-documented here.

**Review by flowerysong — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 1 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py]: Done.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 2 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py]: Done.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 3 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py]: Removed.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 5 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py]: Done.

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 6 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py:90]: Done

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 7 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py]: Done

File: lib/ansible/modules/cloud/amazon/lightsail_lb.py (hunk 4 above)
Comment (gencebay) [reply lib/ansible/modules/cloud/amazon/lightsail_lb.py:296]: Do you have any suggestions?

**Comment by gencebay** [comment]
What is status of the PR, is there anything else you need ?

**Comment by Akasurde** [comment]
Hi @gencebay, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.aws<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>

**Pull request closed without merging**